src/internal/_set.js

```diff
@@ -0,0 +1,23 @@
+var _extend = require('./_extend');
+var _has = require('./_has');
+var _isArray = require('./_isArray');
+var isEmpty = require('../isEmpty');
+var keys = require('../keys');
+
+module.exports = function _set(update, original) {
+  if (_isArray(original) || _isArray(update)) {
+    return update;
+  }
+  var originalKey = keys(original);
+  var updateKeys = keys(update);
+  if (isEmpty(originalKey) || isEmpty(updateKeys)) {
+    return update;
+  }
+  var result = _extend({}, original);
```

Comment:
I don't know if there are any ways to make this more efficient by folding this into the iteration below.  It might not be possible, but it might be worth investigating.